### ISSUE TITLE
mincore has been removed from OpenBSD current.

### ISSUE THREAD

**Issue by adamsteen** [Solo5/solo5#322]
Hi Guys

`mincore` has been remove from OpenBSD current[1]

> Antiquated mincore(2) will not be needed and was removed, eliminating an interface that exposed physical machine information unnecessarily.

 i am not sure the best way of fixing/not compiling "hvt_module_dumpcore".

```
/home/asteen/devl/solo5/tenders/hvt/hvt_module_dumpcore.c:195:9: error:
      implicit declaration of function 'mincore' is invalid in C99
      [-Werror,-Wimplicit-function-declaration]
    if (mincore(hvt->mem, hvt->mem_size, mvec) == -1) {
        ^
1 error generated.
```

Before i make a change, i was hoping to get a direction on fixing this.

**gdb** and **dumpcore** are a fair way off from getting OpenBSD support with ocaml/opam porting, and mprotect ept kernel work required for general Mirage use.

Cheers
Adam

[1] https://marc.info/?l=openbsd-cvs&m=154723400730318&w=2

**Comment by mato** [comment]
> mincore has been remove from OpenBSD current[1]

While it's nice that the OpenBSD folks are removing "antiquated" system calls, I don't see any sensible way of implementing dumpcore functionality without a `mincore()` or equivalent. Yes, we could write out the guest memory in its entirety, but that would cause the host to page in all pages of guest memory, not just those it has actually touched.

The OpenBSD commit mentions "only a few pieces of software use it, generally poorly thought out.  they are being fixed, so mincore() can be deleted.". May be worth asking (with a pointer to the Solo5 dumpcore code) how they would recommend fixing this.

Then again, it is a fairly obscure feature, so it could also just not be supported on OpenBSD, up to you.

> Before i make a change, i was hoping to get a direction on fixing this.

Can you wait a week or so? I need to rip out the specialization-at-compile-time in hvt, then it'll be easier to desupport dumpcore for OpenBSD.

**Comment by adamsteen** [comment]
Yeah, my todo list is a mile long, just let me know and I will fix up OpenBSD!

**Comment by mato** [comment]
@adamsteen Will do. In any case I'm going to close #278 for now as it looks like that won't happen given the removal of `mincore()`.

**Comment by aaglenn** [comment]
I hate to be that person, but I just arrived to the MirageOS retreat and fired up solo5 on my 6.5-current snapshot and ran into this. If there are patches or a branch I can help test, Id really love to.

in the mean time I'll just run a 6.4-release in vmm and nest (which I think is possible...)

**Comment by adamsteen** [comment]
@aaglenn

See https://github.com/adamsteen/solo5/tree/openbsd_fix?files=1 for a temporary fix you could pin

Also OpenBSD now has Ocaml 4.07.1 and OPAM 2.0.3 in ports, hopefully packages soon, see https://gist.github.com/adamsteen/6bdae8dc93d8f91f9eb6cf1dbbbbe4b5 for OpenBSD for 6.4 help

With the above branch I am able to run the solo5 tests, but when trying to run mirage_skeleton tutorial noop I get a fatal error, have only just started to look into it, was going to start at 6.4 and work forwards to see when it broke

Hope that gives you something

**Comment by mato** [comment]
@aaglenn @adamsteen The build system changes in #326 will fix this by not compiling any of the hvt debug modules on OpenBSD. If you want to help, it's worth testing the tip of the branch mentioned there with OpenBSD-current but note that it's currently a moving target and does _not_ work with OPAM/MirageOS correctly yet.

**Comment by mbelop** [comment]
fwiw, mincore was removed because of https://arxiv.org/abs/1901.01161

**Comment by adamsteen** [comment]
fixed in #345. `solo-hvt` is built correctly and `solo-hvt-debug` is not built, not causing any issues.

closing.
